DM-42306: Performance optimizations

timj opened the pull request:
Depends on lsst/resources#76

## Checklist

- [x] ran Jenkins
- [x] added a release note for user-visible changes to `doc/changes`

timj committed:
Add a dry_run parameter to Butler.transfer_from

timj committed:
Avoid retrieving the dimension value twice in loop

timj committed:
Only ask for records that the template needs

timj committed:
Add trust flag when creating Location

If the relative path is coming directly from a datastore record
we know it can't be something that is trying to ../ out of the
datastore.

The safety check in the Location constructor has significant
overhead.

Although with client/server this might be less true.

timj committed:
Convert the lookup content to dict from Config

timj committed:
Only merge if there is something to merge

timj committed:
Annotate some ResourcePath constructors that we know are files

This can lead to some small performance improvement when ResourcePath
does not need to check itself whether something is a file or not.

timj committed:
Add Location.clone() method

This makes it easier for code to copy a Location without having
to know if deepcopy should be used.

timj committed:
Now force LocationFactory to be using files in datastore

timj committed:
Speed up template validation

The .fields method was slow and also being called four times.
Now call a grouped fields method once that returns all the answers.

timj committed:
We know that adding butler.yaml is a file URI

timj committed:
Experimental: Thread pool executor

timj committed:
Tweak the exception raised when bad directory is used as file

timj committed:
Ensure that Location.clone() minimizes copying

Add __deepcopy__ and __copy__ special methods that use clone()

timj committed:
Refresh pre-commit

timj committed:
Add news fragments

timj committed:
Fix handling of skypix in file template

timj committed:
Use collections.abc rather than typing

timj committed:
Remove extraneous parens (UP034)

timj committed:
Use type instead of typing.Type (UP006)

timj committed:
Add check for ../ in file template

Important now that we disable check for this in Location constructor.

timj commented on python/lsst/daf/butler/datastores/fileDatastore.py:
```diff
@@ -2449,9 +2485,38 @@ def transfer_from(
         # See if we already have these records
-        target_records = self._get_stored_records_associated_with_refs(refs, ignore_datastore_records=True)
+        if not dry_run:
+            target_records = self._get_stored_records_associated_with_refs(
```
This is the line I'm not sure about. I imagine people expect dry-run to know which records are really needed to be transferred rather than having the system think that it has to do all the checks again because the target doesn't have the datastore entries even when it does.

TallJimbo replied:
Is this here because we're worried that the datasets might be present in the registry database, but not have file artifacts in the datastore and/or datastore records in the database?

I think it ought to be fine to assume that if a ref in the graph is in the registry we can skip it, as if it was inserted into the registry without associated datastore content that was probably intentional.

timj replied:
I got rid of the change. The point was to try to make my profiles take as long as if the files weren't already transferred, but in reality it was a minor blip and it's better if the dry-run implementation does report that all the datasets have already been transferred and so you know that it won't do anything if you run it for real.

TallJimbo commented on python/lsst/daf/butler/_butler.py:
```diff
@@ -1335,6 +1336,10 @@ def transfer_from(
+        dry_run : `bool`, optional
+            If `True` the transfer will be processed without any modifications
+            made to the target butler and as if the target butler did not
+            have any of the datasets.
```
It will check to see whether the artifacts in the graph exist, right?

I agree with your comment on Jira that it'd be better if it didn't assume that the datasets were not in the target butler, but this is useful enough as a starting point.  Maybe move that qualification to a `Notes` section, and amend it to say that this might change in the future?

timj replied:
It does everything except update target butler.

TallJimbo commented on python/lsst/daf/butler/datastores/fileDatastore.py:
```diff
@@ -2490,9 +2491,17 @@ def transfer_from(
+            )
+        else:
+            # In dry-run mode we assume there is nothing in target repo.
+            # (otherwise everything would report that no work is needed.)
```
I don't understand this.  In the typical production case, the file artifacts are mostly all present and the others are completely absent (since we run directly in the datastore root, so there are no file transfers), and the "work needed" is inserting stuff into the database.  What keeps the dry-run from doing the same, and then (doing the equivalent of) rolling back the DB transaction instead of committing it?

timj replied:
Right, I've removed this. The loop now does act like a dry run in that if the dataset is known to the datastore already it doesn't even attempt to do the copy (the copy would never happen because that's masked out later with dry-run). I'm now wondering why this check doesn't happen at the start of the `Datastore.transfer_from()` on the basis that the target datastore thinking it has the files obviates the need to actually check the file system for them.

TallJimbo commented on python/lsst/daf/butler/datastore/file_templates.py:
```diff
@@ -449,7 +449,7 @@ def format(self, ref: DatasetRef) -> str:
         # dimensions to fill out templates).
         fields: dict[str, object] = {
-            k: ref.dataId.get(k) for k in ref.datasetType.dimensions.names if ref.dataId.get(k) is not None
+            k: v for k in ref.datasetType.dimensions.names if (v := ref.dataId.get(k)) is not None
```
```
fields = dict(dataId.mapping)
```
is all you need now.

[DM-21840](https://jira.lsstcorp.org/browse/DM-21840) is still open because `None` values in data IDs are not prohibited by DB constraints yet, but it's long been true that conditions like "exposure with NULL filter" is for all intents and purposes forbidden, because if you do it the exposure becomes unqueryable.

TallJimbo commented on python/lsst/daf/butler/_location.py:
```diff
@@ -49,7 +49,12 @@ class Location:
+        self,
+        datastoreRootUri: None | ResourcePathExpression,
+        path: ResourcePathExpression,
+        _trusted_path: bool = False,
```
`Location` itself is already sufficiently internal that I don't think you need to make this argument extra-private.

TallJimbo reviewed (approved):
Optimizations look great.

I'm not sure I understand how the transfer-from dry-run option works (maybe because I'm not sure precisely how the checking for existing datasets in the target butler works even without dry-run).

I do think we should drop the `ThreadPoolExecutor` commit for now.

timj committed:
Fix updatedFile usage to be more explicit

timj committed:
Revert "Experimental: Thread pool executor"

This reverts commit f212bf5f27a5aac9385ab972821e0a80bdc19876.

We have decided that we are not read for the thread pool speedup.

timj committed:
Allow dry-run to report when it has already transferred something

Dry run is more useful if it really does report what it is going
to do rather than reporting that it might transfer all the files
to the target butler.

timj commented:
> I'm not sure I understand how the transfer-from dry-run option works (maybe because I'm not sure precisely how the checking for existing datasets in the target butler works even without dry-run).

It does all the work that it would have to do to do the transfer but it doesn't copy any files or write to the target butler. I just modified how it works such that it will report that N datasets have already been transferred rather than dry run assuming that it always has to transfer everything.

timj committed:
Get the dimension values directly from the data ID

No longer need to check for None.

timj committed:
Unhide trusted_path parameter
